syntax = "proto3";

package service;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

message Post {
    int32 id = 1;
    string title = 2;
    string description = 3;
    int32 creator_id = 4;
    bool is_private = 5;
    repeated string tags = 6;
    string loyalty_platform = 7;
    google.protobuf.Timestamp created_at = 8;
    google.protobuf.Timestamp updated_at = 9;
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    int32 creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
    string loyalty_platform = 6;
}

message DeletePostRequest {
    int32 id = 1;
    int32 user_id = 2;
}

message UpdatePostRequest {
    int32 id = 1;
    string title = 2;
    string description = 3;
    bool is_private = 4;
    repeated string tags = 5;
    string loyalty_platform = 6;
    int32 user_id = 7;
}

message GetPostRequest {
    int32 id = 1;
    int32 user_id = 2;
}

message ListPostsRequest {
    int32 page = 1;
    int32 limit = 2;
    int32 user_id = 3;
}

message ListPostsResponse {
    repeated Post posts = 1;
    int32 total = 2;
}

service PostsService {
    rpc CreatePost(CreatePostRequest) returns (Post);
    rpc DeletePost(DeletePostRequest) returns (google.protobuf.Empty);
    rpc UpdatePost(UpdatePostRequest) returns (Post);
    rpc GetPost(GetPostRequest) returns (Post);
    rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
}